lib/postprocessing/image.js

```diff
@@ -0,0 +1,306 @@
+        //       would like to avoid (extra time). for now we trust the worker
+        return dimension.width === instructions.width;
+
+    } else if (instructions.height) {
+        // TODO: this is an optimistic check omitting validating width is in the right aspect ratio.
+        //       doing this would requiring reading the original source file dimensions which we
+        //       would like to avoid (extra time). for now we trust the worker
+        return dimension.height === instructions.height;
+    }
+
+    // if no width or height is set, we assume nothing to be done,
+    // as rendition should be original size which we do not know here
+    return true;
+}
+
+function hasJpegQuality(metadata, instructions) {
+    if (metadata.format === "JPEG" && instructions.quality) {
+        return metadata.quality === instructions.quality;
+    }
+    // if instructions do not request it, we accept any and assume it
+    return true;
+}
+
+async function needsImagePostProcess(rendition) {
+    const instructions = rendition.instructions;
+
+    // verify if the requested format is one we can output from post-processing
+    if (!isSupportedOutputFormat(instructions)) {
+        console.log(`skipping post-processing because output fmt '${instructions.fmt}' is not supported`);
+        return false;
+    }
+
+    await logImageMagickVersion();
+
+    // inspect intermediate rendition
+    const metadata = await getImageMetadata(rendition.path);
+
+    // check if it can read the intermediate rendition
+    if (!isSupportedInputFormat(metadata)) {
+        console.log(`skipping post-processing because intermediate rendition format '${metadata.format}' is not supported`);
+        return false;
+    }
+
+    // check if there is an instruction that definitely needs post processing
+    // fast check first, which does not requiring reading image
+    if (hasExtraInstructions(instructions)) {
+        console.log("running post-processing because instructions require it");
+        return true;
+
+    } else if (!isTargetFormat(metadata, instructions)) {
+        console.log("running post-processing to convert to target format");
+        return true;
+
+    } else if (!isExpectedSize(metadata, instructions)) {
+        console.log("running post-processing to resize");
+        return true;
+
+    } else if (!hasJpegQuality(metadata, instructions)) {
+        console.log("running post-processing to set jpg quality");
+        return true;
+    }
+
+    console.log("skipping post-processing because intermediate rendition meets the target");
+    return false;
+}
+
+// ----------------------------- image conversion -----------------
+
+async function logImageMagickVersion() {
+    const { stdout } = await exec("identify --version");
+    console.log(".....................................................................................................");
+    console.log(stdout.toString().trim());
```

Comment:
add imagemagick version as a metric maybe?

Reply:
I think we don't need to look it up very often, and if we do, seeing it in splunk should be enough...